llvm/lib/Analysis/ValueTracking.cpp

```
@@ -4940,11 +4940,8 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
       // subnormal input could produce a negative zero output.
       const Function *F = II->getFunction();
       if (Q.IIQ.hasNoSignedZeros(II) ||
-          (F && KnownSrc.isKnownNeverLogicalNegZero(*F, II->getType()))) {
+          (F && KnownSrc.isKnownNeverLogicalNegZero(*F, II->getType())))
         Known.knownNot(fcNegZero);
-        if (KnownSrc.isKnownNeverNaN())
```

&& if the source is known positive

---

Isn't it covered by the above code?

https://github.com/llvm/llvm-project/blob/b27eb0ae8280675fc8fb249d39f1ccafa3ee2187/llvm/lib/Analysis/ValueTracking.cpp#L4932-L4937

---

Semi-related, this should be using propagateNaN instead of repeating the snan logic

---

I think we are missing handling of the nnan flag on the sqrt instruction itself

---

Or that just comes from the RAII thing

---

> I think we are missing handling of the nnan flag on the sqrt instruction itself

See https://github.com/llvm/llvm-project/blob/f5c8242042c959cbf476778e1d8a874fc8747801/llvm/include/llvm/Analysis/ValueTracking.h#L510-L527

---

See the test `sqrt_negative_input_nnan`.